### PR TITLE
Refactor `et run` (and friends).

### PR DESCRIPTION
**Pull request by matanlurey**
Does a few things:

- Refactors `run_command_test` significantly to reduce global fixtures
- Replaced stringly-typed things with enum-like objects
- Adds a lot stronger coverage for `run_command` to make future refactors safer
- Takes advantage of `package:test` having a workable matchers system and uses it
- Changes `return 1` into `throw FatalError(...)` where it makes sense in `run_command`

As a result of the refactoring work, I also fixed a bug: Closes https://github.com/flutter/flutter/issues/147646.

### CODE REVIEW

**Comment by matanlurey** [comment]
@johnmccutchan If you are looking to kill a few minutes, would appreciate a review. Thanks!

**Review by johnmccutchan — APPROVED** [review]
I have some nits/questions.

Such a nice cleanup. Thanks for this!

File: tools/engine_tool/lib/src/flutter_tool_interop/device.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/// A representation of the parsed device from the `flutter devices` command.
+///
+/// See <https://github.com/flutter/flutter/blob/9441f9d48fce1d0b425628731dd6ecab8c8b0826/packages/flutter_tools/lib/src/device.dart#L892-L911>.
+@immutable
+final class Device {
+  /// Creates a device with the given [name], [id], and [targetPlatform].
+  const Device({
+    required this.name,
+    required this.id,
+    required this.targetPlatform,
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/engine_tool/lib/src/flutter_tool_interop/device.dart:17]: Let me do this as a follow-up as part of https://github.com/flutter/flutter/issues/155960, I don't like adding code that will go unused.

File: tools/engine_tool/lib/src/commands/run_command.dart [hunk 2]
```diff
@@ -21,6 +25,7 @@ final class RunCommand extends CommandBase {
     required Map<String, BuilderConfig> configs,
     super.help = false,
     super.usageLineLength,
+    @visibleForTesting FlutterTool? flutterTool,
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/engine_tool/lib/src/commands/run_command.dart:28]: Basically this is "don't pass this argument outside of a test", not "the private member is for tests".

File: tools/engine_tool/lib/src/commands/run_command.dart [hunk 3]
```diff
@@ -114,62 +124,60 @@ See `flutter run --help` for a listing
...
-    final RunTarget? target = await _runTarget;
-    final List<Label> buildTargetsForShell =
-        target?.buildTargetsForShell() ?? <Label>[];
+    final target = await _runTarget;
+    final buildTargetsForShell = target?.buildTargetsForShell ?? [];
 
-    final String dashJ = argResults![concurrencyFlag] as String;
-    final int? concurrency = int.tryParse(dashJ);
+    final dashJ = argResults![concurrencyFlag] as String;
+    final concurrency = int.tryParse(dashJ);
```
[the remark this answers is not in the document]
Comment (matanlurey) [reply tools/engine_tool/lib/src/commands/run_command.dart:176]: Done.

File: tools/engine_tool/lib/src/flutter_tool_interop/device.dart [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import 'dart:convert';
```
Comment (zanderso) [inline tools/engine_tool/lib/src/flutter_tool_interop/device.dart:1]: Needs the engine copyright header.

File: tools/engine_tool/lib/src/flutter_tool_interop/flutter_tool.dart [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import 'dart:convert';
```
Comment (zanderso) [inline tools/engine_tool/lib/src/flutter_tool_interop/flutter_tool.dart:1]: ditto.

File: tools/engine_tool/test/src/matchers.dart [hunk 6]
```diff
@@ -0,0 +1,92 @@
+import 'package:engine_tool/src/logger.dart';
```
Comment (zanderso) [inline tools/engine_tool/test/src/matchers.dart:1]: ditto

**Review by zanderso — COMMENTED** [review]
DBC

**Pull request merged**